experimental/DoubleAndHyperComplexes/src/Objects/tensor_products.jl

```diff
@@ -196,3 +196,186 @@ function tensor_product(list::Vector{T}) where {T<:ComplexOfMorphisms}
+  return tensor_product(v...)
+end
+
+# If not, we can wrap it here so that the correct output is produced
+function _tensor_product(v::ModuleFP...)
+  return tensor_product(v...; task=:with_map)
+end
+
+function can_compute(fac::HCTensorProductChainFactory, c::AbsHyperComplex, I::Tuple)
+  i = collect(I)
+
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  return all(k->can_compute_index(fac.factors[k], Tuple(j[k])), 1:length(j))
+end
+
+### Production of the morphisms 
+struct HCTensorProductMapFactory{MorphismType} <: HyperComplexMapFactory{MorphismType}
+  factors::Vector{<:AbsHyperComplex}
+
+  function HCTensorProductMapFactory(
+      ::Type{MT},
+      factors::Vector{<:AbsHyperComplex}
+    ) where {MT}
+    return new{MT}(factors)
+  end
+end
+
+function zero_map(M::ModuleFP, N::ModuleFP)
+  base_ring(M) === base_ring(N) || error("base rings must coincide")
+  return hom(M, N, elem_type(N)[zero(N) for i in 1:ngens(M)], check=false)
+end
+
+function (fac::HCTensorProductMapFactory{MorphismType})(c::AbsHyperComplex, p::Int, I::Tuple) where {MorphismType}
+  next = collect(I)
+  next[p] = next[p] + (direction(c, p) == :chain ? -1 : 1)
+  J = Tuple(next)
+
+  if iszero(c[I]) || iszero(c[J])
+    return zero_map(c[I], c[J])
+  end
+  
+  i = collect(I)
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  maps = MorphismType[]
+  d = length.(j)
+  k = findfirst(k->sum(d[1:k]; init=0) >= p, 1:length(j))
+  k === nothing && error("direction out of bounds")
+  q = p - sum(d[1:k-1]; init = 0)
+  maps = MorphismType[identity_map(fac.factors[l][Tuple(j[l])]) for l in 1:k-1]
+  push!(maps, map(fac.factors[k], q, Tuple(j[k]))) 
+  maps = vcat(maps, MorphismType[identity_map(fac.factors[l][Tuple(j[l])]) for l in k+1:length(j)])
+  return tensor_product(c[I], c[J], maps)
+end
+
+function can_compute(fac::HCTensorProductMapFactory, c::AbsHyperComplex, p::Int, I::Tuple)
+  i = collect(I)
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  d = length.(j)
+  l = findfirst(l->sum(d[1:l]; init=0)>=p, 1:length(j))
+  l === nothing && error("mapping direction out of bounds")
+  return can_compute_map(fac.factors[l], p - sum(d[1:l-1]; init=0), Tuple(j[l]))
+end
+
+### The concrete struct
+@attributes mutable struct HCTensorProductComplex{ChainType, MorphismType} <: AbsHyperComplex{ChainType, MorphismType} 
+  internal_complex::HyperComplex{ChainType, MorphismType}
+  factors::Vector{<:AbsHyperComplex}
+
+  function HCTensorProductComplex(
+      factors::Vector{<:AbsHyperComplex}
+    )
+    CT = reduce(typejoin, chain_type.(factors))
+    MT = reduce(typejoin, morphism_type.(factors))
```

Comment:
Note that this first allocates a vector to store the result of `chain_type.(factors)`. This can be avoided by e.g. using `mapreduce`
```suggestion
    CT = mapreduce(chain_type, typejoin, factors)
    MT = mapreduce(morphism_type, typejoin, factors)
```

Reply:
Beyond that, I think usage of `typejoin` warrants a comment justifying its use... I.e. the big picture "why" it is used.

Otherwise I would have no idea how to debug this code if I run into it again in 2 years.